Explicit name for doc snippets

probakowski opened the pull request:
This change adds option to specify explicit name for doc snippet.
This name will be used instead of line number when creating
yml file in buildRestTests task.
Stable names should improve tracking changes through history and allow
Gradle to skip tests on non-code docs changes.

probakowski committed:
Explicit name for doc snippets

This change adds option to specify explicit name for doc snippet.
This name will be used instead of line number when creating
yml file in buildRestTests task.
Stable names should improve tracking changes through history and allow
Gradle to skip tests on non-code docs changes.

elasticmachine commented:
Pinging @elastic/es-docs

elasticmachine commented:
Pinging @elastic/es-core-infra

probakowski commented:
Only added name to 1 snippet (cat/alias) to avoid work if change is generally not well received

mark-vieira commented:
This is awesome. Once we go back and add explicit names to all snippets I wonder if this is something we could then enforce as a precommit check.

alpar-t reviewed (changes requested):
@nik9000 what's your opinion on this ? 

I an see why @mark-vieira considers this awesome, and I agree it's effects on caching are indeed awesome, but I'm not fully convinced we want to achieve this in this way. 
My main concern is that this doesn't really add any value to the documentation, most of the time people will struggle to assign names, will come up with non meaningful ones, or just copy paste the same name ( do we have some protection against that ). 

Could we use the line from before the snippet as a name ? 
`Hidden setup for example:`  in this PR to achieve the same benefits ? 

Or maybe assign a number to the snippets ? I realize that the numbering will still change when adding removing. 

Finally could we maybe track the line number of some form of meta data ? 
Not sure how it would be implemented, but the idea would be that we can show a line number in the failure without it influencing caching but still preserving the ability to quickly jump to it's source code.

mark-vieira commented:
@atorok brings up some good points. We don't want to sacrifice usability and ease of authoring for this. There are other ways we can implement better cacheability for these tests, such as doing a post-processing step to normalize out these test names. If we think that having to assign static names is too cumbersome then we can go another direction.

probakowski committed:
Avoid duplication in names

nik9000 commented:
> @nik9000 what's your opinion on this ?

I certainly don't mind naming the snippets if it makes your life simpler, though I'd try and emulate [Asciidoctor](https://asciidoctor.org/docs/user-manual/#anchordef)'s syntax for adding `id`s to the the snippet, I think.

If you just need a stable identifier why not hash it? We use murmurhash3 in the docs build for other purposes and that seems to work well.

probakowski commented:
I agree that naming is usually hard, but for examples in test in my opinion it's not. You add example to show a specific thing, and names don't need to be very descriptive or in any particular format. Either one word or whole sentence would do. 
The biggest win for me is that you have explicit anchor in source that can easily be referenced. You can search for it in no time, even without opening that file which makes connecting for example failure to particular test a breeze. It can be also used in different ways - e.g. you could create link to particular example similarly as you can to section header currently with small change to doc generator.
I don't really like the idea of using previous sentence as name because this tells you nothing and this can be very lon. It would be easy to search for though (at least if you do no post-processing on it), I grant you that.
We can use section header with something like 
```java
matcher = line =~ /={2,} (.*)/
if(matcher.matches()) {
    name = matcher.group(1)
    return
}
matcher = line =~ /\[\[(.*)]]/
if(matcher.matches()) {
    name = matcher.group(1)
    return
}
```
but there are cases where there are up to 7 or 8 examples in single section, so we would end up (with new code to avoid duplications) with names like enabled_1, enabled_2, etc. Not terrible but harder to pinpoint exact location of failure.

@nik9000 problem with hash as test name is that you can't easily connect test name with code location. We would need additional metadata in separate file with hash->line number mapping which adds complexity to debugging process.

probakowski commented:
@nik9000 I think part of my code above (```matcher = line =~ /\[\[(.*)]]/```) would catch anchors from asciidoctor, so yeah, probably we can just use those as snippets names

nik9000 commented:
> I don't really like the idea of using previous sentence as name because this tells you nothing and this can be very lon.

++



> @nik9000 problem with hash as test name is that you can't easily connect test name with code location. We would need additional metadata in separate file with hash->line number mapping which adds complexity to debugging process.

++

> @nik9000 I think part of my proposal above (`matcher = line =~ /\[\[(.*)]]/`) would catch anchors from asciidoctor, so yeah, probably we can just use those as snippets names to avoid new custom syntax

If naming things is really helpful why not name everything? Instead of falling back to the header always pick up the name. Asciidoctor has three ways of setting an ID on a thing, but I think you may as well stick to just one. Try, like,
```
[source,console,id=bucket-script-agg-context-example]
```

That way you don't need a new line.

One wrinkle: ids in asciidoctor are global. They also generate empty anchor tags on the page that let you link to the specific example directly. IDs being global is a pain. The anchor tags is a net positive though.

probakowski committed:
Changes id declaration, more examples

probakowski committed:
Fix names in examples

probakowski commented:
@nik9000 I like `[source,console,id=bucket-script-agg-context-example]`, I've adjusted code to use that one. 
I've added also more examples and fix for duplicated names.
@atorok @mark-vieira what do you think?

atorok commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/doc/SnippetsTask.groovy:
```diff
@@ -153,19 +154,21 @@ public class SnippetsTask extends DefaultTask {
                     return
                 }
-                matcher = line =~ /\["?source"?,\s*"?([-\w]+)"?(,.*)?].*/
+                matcher = line =~ /\["?source"?,\s*"?([-\w]+)"?(,((?!id=).)*(id="?([-\w]+)"?)?(.*))?].*/
```
Could we add a unit test for this ?

probakowski replied:
Unit test added

atorok commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/doc/RestTestsFromSnippetsTask.groovy:
```diff
@@ -235,7 +237,13 @@ public class RestTestsFromSnippetsTask extends SnippetsTask {
                 current.println('---')
-                current.println("\"line_$test.start\":")
+                if (test.name != null && test.name.isBlank() == false) {
+                    def num = names.compute(test.name, { key, old -> old == null ? 1 : old + 1 })
```
I'm not 100% sure we want to allow this. 
Any reason they shouldn't all be unique so we can just err here if we detect duplicates ?

probakowski replied:
I was unsure as well. I've changed it to exception.

nik9000 replied:
If the ids aren't unique then the docs won't build either1

probakowski committed:
Unit test added

probakowski committed:
Throw exception on duplicate names

probakowski commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into stable-snippets-names

atorok commented on buildSrc/src/test/groovy/org/elasticsearch/gradle/doc/SnippetsTaskTests.groovy:
```diff
@@ -0,0 +1,43 @@
+
+import org.elasticsearch.gradle.test.GradleUnitTestCase
+
+class SnippetsTaskTests extends GradleUnitTestCase {
```
Please add a Java test. We are looking to replace groovy all new classes should be Java.

probakowski replied:
Done

probakowski committed:
Moved UT to Java

probakowski committed:
Merge branch 'master' into stable-snippets-names
